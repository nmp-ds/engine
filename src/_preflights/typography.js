import { entriesToCss } from '@unocss/core';

const fontSizeBaseValues = {
  '--w-font-size-xs': "1.2rem",
  '--w-font-size-s': "1.4rem",
  '--w-font-size-m': "1.6rem",
  '--w-font-size-ml': "2rem",
  '--w-font-size-l': "2.2rem",
  '--w-font-size-xl': "2.8rem",
  '--w-font-size-xxl': "3.4rem",
  '--w-font-size-xxxl': "4.8rem",
};

const lineHeightBaseValues = {
  '--w-line-height-xs': "1.6rem",
  '--w-line-height-s': "1.8rem",
  '--w-line-height-m': "2.2rem",
  '--w-line-height-ml': "2.6rem",
  '--w-line-height-l': "2.8rem",
  '--w-line-height-xl': "3.4rem",
  '--w-line-height-xxl': "4.1rem",
  '--w-line-height-xxxl': "5.6rem",
};

const headingsReset = `
  h1  {
    font-size: var(--w-font-size-xxl);
    line-height:var(--w-line-height-xxl);
  }

  h2  {
    font-size:var(--w-font-size-xl);
    line-height:var(--w-line-height-xl);
  }

  h3  {
    font-size:var(--w-font-size-l);
    line-height:var(--w-line-height-l);
  }

  h4  {
    font-size:var(--w-font-size-m);
    line-height:var(--w-line-height-m);
  }

  h5  {
    font-size:var(--w-font-size-s);
    line-height:var(--w-line-height-s);
  }
`;

export const typographyBase = {
  layer: 'preflights',
  getCSS() {
    const css = entriesToCss(Object.entries({ ...fontSizeBaseValues, ...lineHeightBaseValues }));
    return `${css}${headingsReset}`;
  },
};
