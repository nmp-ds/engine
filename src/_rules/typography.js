import { textMap, lineHeightMap } from '#utils';

export const typography = [
  [/^text-(12|14|16|20|22|28|34|48)$/, ([, d]) => ({ 'font-size': `var(--w-font-size-${textMap[d]})`, 'line-height': `var(--w-line-height-${textMap[d]})` })],
  [/^text-(xs|s|m|ml|l|xl|xxl|xxxl)$/, ([, size]) =>
    ({ 'font-size': `var(--w-font-size-${size})`, 'line-height': `var(--w-line-height-${size})` }),
  ],
  [/^leading-(16|18|22|26|28|34|41|56)$/, ([, d]) => ({ 'line-height': `var(--w-line-height-${lineHeightMap[d]})` })],
  [/^leading-(xs|s|m|ml|l|xl|xxl|xxxl)$/, ([, size]) =>
    ({ 'line-height': `var(--w-line-height-${size})` }),
  ],
];
