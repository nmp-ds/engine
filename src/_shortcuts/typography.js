export const typographyAliases = {
  t1: 'text-xxl font-bold',
  t2: 'text-xl font-bold',
  t3: 'text-l font-bold',
  t4: 'text-m font-bold',
  t5: 'text-s font-bold',
  t6: 'text-xs font-bold',
  h1: 't1',
  h2: 't2',
  h3: 't3',
  h4: 't4',
  h5: 't5',
  'text-display': 'text-xxxl',
  'text-preamble': 'text-ml',
  'text-body': 'text-m',
  'text-caption': 'text-s',
  'text-detail': 'text-xs',
};