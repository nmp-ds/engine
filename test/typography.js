import { expect, test } from 'vitest';
import { setup } from './_helpers.js';
import { textMap, lineHeightMap } from '#utils';
import { typographyAliases } from '../src/_shortcuts/typography.js';

setup();

test('typography - text classes ', async ({ uno }) => {
  const classes = Object.entries(textMap).map(([number, size]) => ([`text-${number}`, `text-${size}`])).flat(1);
  const { css } = await uno.generate(classes);
  expect(css).toMatchSnapshot();
});

test('typography - leading classes ', async ({ uno }) => {
  const classes = Object.entries(lineHeightMap).map(([number, size]) => ([`leading-${number}`, `leading-${size}`])).flat(1);
  const { css } = await uno.generate(classes);
  expect(css).toMatchSnapshot();
});

test('typography - leading classes with arbitrary values', async ({ uno }) => {
  const classes = [`leading-[24]`, `leading-[11]`, `leading-[24px]`, `leading-[24rem]`].flat();
  const { css } = await uno.generate(classes);
  expect(css).toMatchSnapshot();
});

test('shortcuts', async ({ uno }) => {
  const classes = Object.keys(typographyAliases);
  const { css } = await uno.generate(classes);
  expect(css).toMatchSnapshot();
});
