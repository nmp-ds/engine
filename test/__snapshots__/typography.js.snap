// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`shortcuts 1`] = `
"/* layer: shortcuts */
.h1,
.t1{font-weight:700;font-size:var(--w-font-size-xxl);line-height:var(--w-line-height-xxl);}
.h2,
.t2{font-weight:700;font-size:var(--w-font-size-xl);line-height:var(--w-line-height-xl);}
.h3,
.t3{font-weight:700;font-size:var(--w-font-size-l);line-height:var(--w-line-height-l);}
.h4,
.t4{font-weight:700;font-size:var(--w-font-size-m);line-height:var(--w-line-height-m);}
.h5,
.t5{font-weight:700;font-size:var(--w-font-size-s);line-height:var(--w-line-height-s);}
.t6{font-weight:700;font-size:var(--w-font-size-xs);line-height:var(--w-line-height-xs);}
.text-body{font-size:var(--w-font-size-m);line-height:var(--w-line-height-m);}
.text-caption{font-size:var(--w-font-size-s);line-height:var(--w-line-height-s);}
.text-detail{font-size:var(--w-font-size-xs);line-height:var(--w-line-height-xs);}
.text-display{font-size:var(--w-font-size-xxxl);line-height:var(--w-line-height-xxxl);}
.text-preamble{font-size:var(--w-font-size-ml);line-height:var(--w-line-height-ml);}"
`;

exports[`typography - leading classes  1`] = `
"/* layer: default */
.leading-16,
.leading-xs{line-height:var(--w-line-height-xs);}
.leading-18,
.leading-s{line-height:var(--w-line-height-s);}
.leading-22,
.leading-m{line-height:var(--w-line-height-m);}
.leading-26,
.leading-ml{line-height:var(--w-line-height-ml);}
.leading-28,
.leading-l{line-height:var(--w-line-height-l);}
.leading-34,
.leading-xl{line-height:var(--w-line-height-xl);}
.leading-41,
.leading-xxl{line-height:var(--w-line-height-xxl);}
.leading-56,
.leading-xxxl{line-height:var(--w-line-height-xxxl);}"
`;

exports[`typography - leading classes with arbitrary values 1`] = `
"/* layer: default */
.leading-\\\\[11\\\\]{line-height:1.1rem;}
.leading-\\\\[24\\\\]{line-height:2.4rem;}
.leading-\\\\[24px\\\\]{line-height:24px;}
.leading-\\\\[24rem\\\\]{line-height:24rem;}"
`;

exports[`typography - text classes  1`] = `
"/* layer: default */
.text-12,
.text-xs{font-size:var(--w-font-size-xs);line-height:var(--w-line-height-xs);}
.text-14,
.text-s{font-size:var(--w-font-size-s);line-height:var(--w-line-height-s);}
.text-16,
.text-m{font-size:var(--w-font-size-m);line-height:var(--w-line-height-m);}
.text-20,
.text-ml{font-size:var(--w-font-size-ml);line-height:var(--w-line-height-ml);}
.text-22,
.text-l{font-size:var(--w-font-size-l);line-height:var(--w-line-height-l);}
.text-28,
.text-xl{font-size:var(--w-font-size-xl);line-height:var(--w-line-height-xl);}
.text-34,
.text-xxl{font-size:var(--w-font-size-xxl);line-height:var(--w-line-height-xxl);}
.text-48,
.text-xxxl{font-size:var(--w-font-size-xxxl);line-height:var(--w-line-height-xxxl);}"
`;
